Clean up contributing documentation

infinisil opened the pull request:
### Motivation

For [RFC 140](https://github.com/NixOS/rfcs/pull/140) (implementation [part 1](https://github.com/NixOS/nixpkgs/pull/237439)) I want to update documentation on how to contribute packages, see [here](https://github.com/NixOS/nixpkgs/pull/237439/files#diff-d9e77bfcb681fae6baf681117c55d2b9119546a2a8727a876a8b497fa69dcd90).

However, after taking a closer look, contributing documentation is currently missing, unorganized, scattered and/or duplicated throughout various documents:
- A [Contributing section](https://github.com/NixOS/nixpkgs/blob/58b1c722f70fff9e967f140f714f19022cc219d1/README.md#contributing) in the top-level `README.md` containing some minimal info about branches
- The [`CONTRIBUTING.md`](https://github.com/NixOS/nixpkgs/blob/58b1c722f70fff9e967f140f714f19022cc219d1/CONTRIBUTING.md) file, which GitHub prominently links to when you open a pull request, containing some duplicated but also updated copies from the contributing section in the manual.
- The [Contributing section](https://github.com/NixOS/nixpkgs/blob/58b1c722f70fff9e967f140f714f19022cc219d1/doc/contributing.md) in the manual ([rendered](https://nixos.org/manual/nixpkgs/stable/#part-contributing)), containing a bunch of general docs, but also a lot of docs specific to `pkgs`. In fact that section sometimes assumes everybody just wants to make changes to `pkgs`
- Various sections in the Nixpkgs manual for contributing to specific parts, like [this one](https://github.com/NixOS/nixpkgs/blob/58b1c722f70fff9e967f140f714f19022cc219d1/doc/languages-frameworks/python.section.md#contributing-contributing) for Python (which has taken the `#contributing` anchor..), or [this one](https://github.com/NixOS/nixpkgs/blob/58b1c722f70fff9e967f140f714f19022cc219d1/doc/languages-frameworks/r.section.md#updating-the-package-set-updating-the-package-set) for R
- Various documents throughout the source code not included in the manual, like [this one](https://github.com/NixOS/nixpkgs/blob/58b1c722f70fff9e967f140f714f19022cc219d1/pkgs/development/haskell-modules/HACKING.md) for Haskell, or [here](https://github.com/NixOS/nixpkgs/blob/58b1c722f70fff9e967f140f714f19022cc219d1/pkgs/servers/nextcloud/packages/README.md) for nextcloud.

Instead of trying to shoehorn the documentation for RFC 140 into this mess, it would be better to first clean up the overall structure. Part of this was already discussed in https://github.com/NixOS/nixpkgs/issues/244056, and this PR will attempt to implement this roughly.

### Goal

The following organization is proposed:
- `README.md`: A non-technical contribution introduction, links to `CONTRIBUTING.md`
- `CONTRIBUTING.md`: General contribution guidelines that apply to all of Nixpkgs, such as:
  - The file structure of Nixpkgs, linking to more detailed `README.md`'s for individual parts 
  - Different ways of contributing: Proposing changes, Reviewing changes, Merging changes
  - Flow of changes from PR merge to release (including staging)
  - Coding Conventions (syntax, formatting, filenames)
- `**/README.md`: Contributing documentation specific to other parts of Nixpkgs, also linking to `CONTRIBUTING.md`. This should include how to modify, test and document each part of the code base.
  - `pkgs/README.md`: Contribution documentation for packages (creating, updating, testing packages, etc.)
  - `doc/README.md`: Contribution documentation for the user manual
  - `lib/README.md`: Contributing to `lib`, already done with https://github.com/NixOS/nixpkgs/pull/244044
  - ...
- Nixpkgs manual: No contributing docs anymore, only link to `CONTRIBUTING.md` and only include user-level documentation:
  - How to report bugs or request a change (using the issue tracker)
  - How to test pull requests (without reviewing the code)
  - How to check if merged pull requests are available in releases

This PR will attempt to set the baseline for this by at least moving the existing documentation to the right place and making it work together.

Best reviewed commit-by-commit!

### Todo

- [x] Define structure of `CONTRIBUTING.md`
- [x] Roughly move all docs relevant to all of Nixpkgs to `CONTRIBUTING.md`
- [x] Move all docs relevant to `pkgs` into a new `pkgs/README.md`
- [x] Clean up and reconcile moved documentation
  - I couldn't clean up _everything_, but I made a good effort for at least `CONTRIBUTING.md`
- [ ] Ensure everything is cross-linked nicely
- [x] Ensure that removed anchors in the manual keep working
- [x] Add myself as a code owner to all relevant files
  - See also https://github.com/NixOS/nixpkgs/pull/248166

infinisil committed:
doc/contributing: Remove unreferenced legacy diagram

Not used anymore after b3fa79bb8965f7d3353daf99b137170314899a53

RaitoBezarius commented on CONTRIBUTING.md:
````diff
@@ -144,7 +409,14 @@ for review, which allows you to sidestep this issue.
 ```
 
-## Backporting changes
+### (optional) Backporting
````
No, backporting is mandatory for security updates.

RaitoBezarius replied:
(and it's *expected* from maintainers to do due diligence regarding this.)

RaitoBezarius commented on CONTRIBUTING.md:
```diff
@@ -61,16 +239,103 @@ Pull requests should not be squash merged in order to keep complete commit messa
+
+(pushing, creating a pull request, select the right branch (link to above section))
+
+When pull requests are made, our tooling automation bot,
```
Can you add considerations on mass ping & the fact that a PR should be open draft if it does not meet the minimal requirements of having a descriptive title and description? Apparently it's not clear.

infinisil replied:
In the first iteration of this PR I'm just going to move existing documentation without writing new one, and I want to only limit new docs to holes that need to be filled, otherwise this would quickly get too big. I think this here in particular should be out of scope for this initial PR.

RaitoBezarius replied:
In that case, let's not write more things like the `(optional) ` ? :D

infinisil replied:
I do count that as organization, it shows up in the ToC :)

infinisil replied:
Oh but yes I will have to make sure the titles are accurate, so the `(optional)` should be revised for sure

RaitoBezarius replied:
Move + changing the appropriate titles makes it hard to review what changed in general :P.
I'd appreciate if there was a commit for all the title renaming stuff or "what is new" compared to the old contents.

infinisil replied:
Sounds like a good idea, I'll consider that when cleaning up the history

infinisil replied:
I cleaned up the history now. Content was strictly only moved, only adding new titles when really necessary, and then wrapping them with () to indicate they need to be considered. For every commit only contents from a single file was moved.

infinisil committed:
Create unified markdown files for contributing docs

infinisil committed:
README.md: Rough move to new contribution doc files

No content was changed, new titles are wrapped with () to signal that
they will need to be decided on in a future commit.

infinisil committed:
CONTRIBUTING.md: Rough move to new contribution doc files

No content was changed, new titles are wrapped with () to signal that
they will need to be decided on in a future commit.

infinisil committed:
doc/coding-conventions: Rough move to new contribution doc files

No content was changed, new titles are wrapped with () to signal that
they will need to be decided on in a future commit.

Section in the manual have been preserved with a simple redirect to
GitHub, the proper anchors should be filled out in a future commit once
the new section names are decided.

infinisil committed:
doc/contributing-to-documentation: Rough move to new contribution doc files

Section in the manual have been preserved with a simple redirect to
GitHub, the proper anchors should be filled out in a future commit once
the new section names are decided.

infinisil committed:
doc/quick-start: Rough move to new contribution doc files

Section in the manual have been preserved with a simple redirect to
GitHub, the proper anchors should be filled out in a future commit once
the new section names are decided.

infinisil committed:
doc/reviewing-contributions: Rough move to new contribution doc files

No content was changed, new titles are wrapped with () to signal that
they will need to be decided on in a future commit.

Section in the manual have been preserved with a simple redirect to
GitHub, the proper anchors should be filled out in a future commit once
the new section names are decided.

infinisil committed:
doc/submitting-changes: Rough move to new contribution doc files

No content was changed, new titles are wrapped with () to signal that
they will need to be decided on in a future commit.

Section in the manual have been preserved with a simple redirect to
GitHub, the proper anchors should be filled out in a future commit once
the new section names are decided.

infinisil committed:
doc/vulnerability-roundup: Rough move to new contribution doc files

No content was changed, new titles are wrapped with () to signal that
they will need to be decided on in a future commit.

Section in the manual have been preserved with a simple redirect to
GitHub, the proper anchors should be filled out in a future commit once
the new section names are decided.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-07-27-documentation-team-meeting-notes-67/30998/1

infinisil committed:
doc: Add empty development section

infinisil committed:
CONTRIBUTING.md: Move opening issues section to Nixpkgs manual

This isn't relevant to contributors specifically, it applies to all
users

infinisil committed:
CONTRIBUTING.md: Remove temporarily

So that the CONTRIBUTING-new.md can take its place and Git detect the
rename

infinisil committed:
Move CONTRIBUTING-new.md to CONTRIBUTING.md

The -new version was assembled in the previous commits

infinisil committed:
CONTRIBUTING.md: Add intro

infinisil committed:
CONTRIBUTING.md: Add overview

infinisil committed:
pkgs/README.md: Add overview

infinisil committed:
CONTRIBUTING.md: Move pull request template section to the bottom

infinisil committed:
CONTRIBUTING.md: Collect paragraphs into a new commit conventions section

infinisil committed:
CONTRIBUTING.md: Write a new section on how to propose a change

Removing paragraphs from the other sections that aren't necessary anymore

infinisil committed:
CONTRIBUTING.md: Coding conventions -> code conventions

It's not conventions about how to do coding, it's about what the code
should look like

infinisil committed:
CONTRIBUTING.md: Clean up "Making patches" section

Moving and collecting bits into the appropriate places, also writing
some new sentences

infinisil committed:
CONTRIBUTING.md: Move flow of changes section up

infinisil committed:
CONTRIBUTING.md: Replace staging workflow diagram

Can't be rendered in GitHub markdown, replacing it with a Git branch diagram using mermaid

infinisil committed:
CONTRIBUTING.md: Clean up and condense staging workflow section

infinisil committed:
CONTRIBUTING.md: Move release sections together

Will be cleaned up in the next commit

infinisil committed:
CONTRIBUTING.md: Clean up releases section

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-08-08-nixpkgs-architecture-team-meeting-42/31454/1

matthiasbeyer commented:
Feel free to cherry-pick https://github.com/NixOS/nixpkgs/pull/248013 onto this PR and close it. :+1:

infinisil commented:
@matthiasbeyer https://github.com/NixOS/nixpkgs/pull/248013#issuecomment-1671881712

infinisil committed:
CONTRIBUTING.md: Minor formatting fixes

infinisil committed:
CONTRIBUTING.md: Remove duplicated section

How to force push to PR's is already explained in the
"How to propose a change" section at the beginning

infinisil committed:
CONTRIBUTING.md: Move rebasing section up

Right into the "How to propose a change" section, because that's where
it's relevant

infinisil committed:
CONTRIBUTING.md: New section on which branch to use

infinisil committed:
CONTRIBUTING.md: Minor fixes

infinisil committed:
CONTRIBUTING.md: Move sections around

Yeah I'm totally getting lazy with the commits, but it's really hard and
takes a long time..

infinisil committed:
pkgs/README.md: Minor header and link fixes

infinisil committed:
doc/README.md: Cleanup

infinisil committed:
nixos/README.md: GitHub markdown fixes

infinisil committed:
maintainer/README.md: GitHub markdown fixes

infinisil committed:
pkgs/README.md: Remove some mediocre duplicated sentences

Also mainly covered in the section below

infinisil commented:
Okay so I worked a whole bunch more on this and think it's _mostly_ ready. I made a decent effort at trying to keep the Git log bearable, though some commits at the end got rather messy.

And yes, the contributor documentation is _still_ messy even after all of this work, but it's _better_ than before. Especially the `CONTRIBUTING.md` document is like almost readable and makes sense now, that's where I made the most changes. I will highlight the sections I wrote/rewrote for reviewers. Don't bother with everything else, that was just moved.

infinisil commented on CONTRIBUTING.md:
````diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
 ```
 
-## Backporting changes
+## How to backport pull requests
````
This section and subsections were partially rewritten from originally duplicate information.

infinisil commented on CONTRIBUTING.md:
```diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
+
+In a case a contributor definitively leaves the Nix community, they should create an issue or post on [Discourse](https://discourse.nixos.org) with references of packages and modules they maintain so the maintainership can be taken over by other contributors.
+
+# Flow of merged pull requests
```
This section is entirely new

infinisil commented on CONTRIBUTING.md:
```diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
+
+See [this section][branch] to know when to use the release branches.
+
+## Staging
```
This section is entirely rewritten, combining a lot of duplicate information into a smaller footprint

infinisil commented on CONTRIBUTING.md:
```diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
+
+# Conventions
+
+## Branch conventions
```
This section is entirely new and was sorely missing before

infinisil commented on CONTRIBUTING.md:
```diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
+
+See [this section][staging] for more details about such changes propagate between the branches.
+
+### Changes acceptable for releases
```
This section was titled differently before, but I also rewrote its contents to make more sense

infinisil commented on doc/README.md:
```diff
@@ -1,12 +1,115 @@
+
+If you want to link to a man page, you can use `` {manpage}`nix.conf(5)` ``. The references will turn into links when a mapping exists in [`doc/manpage-urls.json`](./manpage-urls.json).
+
+A few markups for other kinds of literals are also available:
+
+- `` {command}`rm -rfi` ``
```
Unfortunately I had to remove the "which is rendered as" parts because GitHub's markdown doesn't support all of the syntax, here is before: https://nixos.org/manual/nixpkgs/stable/#chap-contributing. That's unfortunate but oh well, it's not a big problem in the grand scheme of things :)

infinisil commented on doc/development.md:
```diff
@@ -0,0 +1,10 @@
+# Development of Nixpkgs {#part-development}
```
This is a new manual section intended to contain development information relevant to _users_, such as:

- How to report bugs or request a change (using the issue tracker)
- How to test pull requests (without reviewing the code)
- How to check if merged pull requests are available in releases

But the section isn't there yet, for now it just contains a starting blurb and then some moved documentation from before.

infinisil commented on pkgs/README.md:
```diff
@@ -0,0 +1,837 @@
+# Contributing to Nixpkgs packages
+
+This document is for people wanting to contribute specifically to the package collection in Nixpkgs.
+See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general information.
+
+## Overview
```
These sections are all new

infinisil reviewed (commented):
These are the sections that I touched most, everything else was mostly just moved.

infinisil committed:
CODEOWNERS: Add myself to the newly added files

matthiasbeyer commented on maintainers/README.md:
```diff
@@ -0,0 +1,114 @@
+When adding users to [`maintainer-list.nix`](./maintainer-list.nix), the following
+checks should be performed:
+
+- If the user has specified a GPG key, verify that the commit is
```
Nowadays it is possible to use ssh keys for commit signing.
There's nothing written here about how to add that information to the maintainers list.

zimbatm replied:
Good point for a follow-up PR. AFAIK the maintainers-list.nix schema doesn´t support it and would have to be extended as well.

infinisil replied:
Yeah, this section was just moved from https://nixos.org/manual/nixpkgs/stable/#reviewing-contributions-individual-maintainer-list, I don't want to update it just yet, but that's a good future change.

Note that there is also some documentation in the `maintainers-list.nix` file itself.

FRidh commented:
Nice work! Good to see this happening.

> * Various sections in the Nixpkgs manual for contributing to specific parts, like [this one](https://github.com/NixOS/nixpkgs/blob/58b1c722f70fff9e967f140f714f19022cc219d1/doc/languages-frameworks/python.section.md#contributing-contributing) for Python (which has taken the `#contributing` anchor..), or [this one](https://github.com/NixOS/nixpkgs/blob/58b1c722f70fff9e967f140f714f19022cc219d1/doc/languages-frameworks/r.section.md#updating-the-package-set-updating-the-package-set) for R

As an example, where would the Python-specific contributing part go to? `**/README.md`? And how is that to be discovered then by say somebody new that wants to contribute? I suggest having a section in `CONTRIBUTING.md`, like the Nixpkgs manual, that covers the larger/common languages/frameworks, essentially just linking to the corresponding `README.md`.

Maybe after that we could use a tool (e.g. sphinx) to compile the various `README.md` into a contributing manual.

zimbatm reviewed (approved):
This is moving in the right direction and is clearly better than what we had before.

Users are now able to navigate the repository and find out about the related documentation, instead of going to a hidden nixpkgs manual URL.

FRidh commented on CONTRIBUTING.md:
```diff
@@ -1,77 +1,193 @@
+
+When sandbox builds are enabled, Nix will setup an isolated environment for each build process. It is used to remove further hidden dependencies set by the build environment to improve reproducibility. This includes access to the network during the build outside of `fetch*` functions and files outside the Nix store. Depending on the operating system access to other resources are blocked as well ( ...
+
+Sandboxing is not enabled by default in Nix due to a small performance hit on each build. In pull requests for [nixpkgs](https://github.com/NixOS/nixpkgs/) people are asked to test builds with sandboxing enabled (see `Tested using sandboxing` in the pull request template) because in [Hydra](https://nixos.org/hydra/) sandboxing is also used.
```
Is the performance hit the reason? I thought it was because the sandbox is not available on all platforms.

infinisil replied:
This is just moved documentation, it was previously in https://nixos.org/manual/nixpkgs/stable/#submitting-changes-tested-with-sandbox, we shouldn't discuss changes to it here.

FRidh commented on CONTRIBUTING.md:
````diff
@@ -1,77 +1,193 @@
 
-Package version upgrades usually allow for simpler commit messages, including attribute name, old and new version, as well as a reference to the relevant release notes/changelog. Every once in a while a package upgrade requires more extensive changes, and that subsequently warrants a more verbose message.
+```ShellSession
+nix run nixpkgs#nixpkgs-review -- pr 12345
````
I thought we kept flakes out of the official documentation in Nixpkgs.

FRidh commented on CONTRIBUTING.md:
```diff
@@ -1,77 +1,193 @@
+#### Meets Nixpkgs contribution standards
 
-## Rebasing between branches (i.e. from master to staging)
+The last checkbox is fits the guidelines in this `CONTRIBUTING.md` file. The contributing document has detailed information on standards the Nix community has for commit messages, reviews, licensing of contributions you make to the project, etc... Everyone should read and understand the standards the community has for contributing before submitting a pull request.
```
This first sentence seems wrong.

infinisil replied:
Good catch, fixed in c99222dbc739b631236f39d570639f9de4482647

FRidh commented on CONTRIBUTING.md:
```diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
+  > **Warning**
+  > Do not use the `nixos-YY.MM` branch, that is a branch pointing to the tested release channel commit
+
+- Instead of making manually making and committing the changes, use [`git cherry-pick -x`](https://git-scm.com/docs/git-cherry-pick) for each commit from the pull request you'd like to backport.
```
```suggestion
- Instead of manually making and committing the changes, use [`git cherry-pick -x`](https://git-scm.com/docs/git-cherry-pick) for each commit from the pull request you'd like to backport.
```

FRidh commented on CONTRIBUTING.md:
```diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
+
+  > **Warning**
+  > Ensure the commits exists on the master branch.
+  > In the case of squashed or rebased merges, the commit hash will change and the new commits can be found in the merge message at the bottom of the master pull request.
```
Good to see this mentioned. It happened often PR's are opened against master/staging as well as with the backport before the change targetting master has been merged.

FRidh commented on CONTRIBUTING.md:
```diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
+
+Package version upgrades usually allow for simpler commit messages, including attribute name, old and new version, as well as a reference to the relevant release notes/changelog. Every once in a while a package upgrade requires more extensive changes, and that subsequently warrants a more verbose message.
+
+Pull requests should not be squash merged in order to keep complete commit messages and GPG signatures intact and must not be when the change doesn't make sense as a single commit.
```
Squash merges merge the commit messages as well, not? And if the merger has a signing key configured, the new commit will be committed with that key. Of course the old ones get lost then. As far as I know there never was any agreement on this part.

matthiasbeyer replied:
> And if the merger has a signing key configured, the new commit will be committed with that key

IIRC that's not true. Github would need access to that private key, wouldn't it? AFAIK it is only committed using the github signing key!

FRidh replied:
Right!

matthiasbeyer replied:
So effectively, all trust is removed in case of a squash-merge/rebase-merge because having a github signature does not require anything except a github account. At least for me, having github signing a commit is equal to not having a signature at all. YMMV of course.

FRidh replied:
>  YMMV of course.

Exactly. Since commit signing has historically been a divisive issue I think we should not mention this requirement it in the contributing guidelines, at least not in this PR. A follow-up PR could cover that.

infinisil replied:
Yeah, this sentence was previously also [in CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/25c37880ecac49691bfa4b5dcb3fc89ad80cb5b7/CONTRIBUTING.md#writing-good-commit-messages) so I don't think we need to discuss it here.

FRidh commented on CONTRIBUTING.md:
```diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
+
+### Release notes
+
+If you removed pkgs or made some major NixOS changes, write about it in the release notes for the next stable release in [`nixos/doc/manual/release-notes`](./nixos/doc/manual/release-notes).
```
```suggestion
If you removed packages or made some major NixOS changes, write about it in the release notes for the next stable release in [`nixos/doc/manual/release-notes`](./nixos/doc/manual/release-notes).
```

FRidh commented on doc/contributing/coding-conventions.chapter.md:
````diff
@@ -1,693 +1,63 @@
-    assert system == "i686-linux";
-      stdenv.mkDerivation { ...
-  ```
+This section has been moved to [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
````
These types of links, while useful when browsing through a browser, now hardcode the branch. I think you can just use relative links.

infinisil replied:
We can't use relative links because this is the rendered manual to be shown on `nixos.org` where the source isn't available.

minijackson commented on CONTRIBUTING.md:
```diff
@@ -368,54 +447,101 @@ The staging workflow is used for all main branches, `master` and `release-YY.MM`
+[commit-conventions]: #commit-conventions
 
-It is possible for community members that have enough knowledge and experience on a special topic to contribute by merging pull requests.
+- When changing the bootloader installation process, extra care must be taken. Grub installations cannot be rolled back, hence changes may break people’s installations forever. For any non-trivial change to the bootloader please file a PR asking for review, especially from \@edolstra.
```
I think it's a bit weird to have this point in the "Commit conventions" section.

infinisil replied:
Oh yeah, I initially wasn't sure where it would fit, but now I'm realizing that `nixos/README.md` would be great: f22e4a3106260cdcc48f41f8f4cb483d9520a39e (of course, that document also desperately needs more content and a cleanup, but that's future work)

minijackson commented on pkgs/README.md:
```diff
@@ -0,0 +1,837 @@
+- Run package-internal tests with `nix-build --attr pkgs.PACKAGE.passthru.tests`
+- Run [NixOS tests](https://nixos.org/manual/nixos/unstable/#sec-nixos-tests) with `nix-build --attr nixosTest.NAME`, where `NAME` is the name of the test listed in `nixos/tests/all-tests.nix`
+- Run [global package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests) with `nix-build --attr tests.PACKAGE`, where `PACKAGE` is the name of the test listed in `pkgs/test/default.nix`
+- See `lib/tests/NAME.nix` for instructions on running specific library tests
```
It may be a good place to mention `nixpkgs-review` again, for completeness' sake

infinisil replied:
Good point, I'd like to defer that to the future when I get to cleaning up the pull request template and the its corresponding section

minijackson reviewed (commented):
Great stuff, thanks a lot!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-08-14-documentation-team-meeting-notes-72/31722/1

infinisil commented:
> I suggest having a section in `CONTRIBUTING.md`, like the Nixpkgs manual, that covers the larger/common languages/frameworks, essentially just linking to the corresponding `README.md`.

@FRidh This would fit into the new `pkgs/README.md`, but that can be in a future change.

> Maybe after that we could use a tool (e.g. sphinx or whatever we use now for the manual) to compile the various README.md into a contributing manual.

I like the idea! It's a bit ironic to move away from rendering the docs ourselves only to render them ourselves again later, but it does make sense looking in the whole :P

infinisil committed:
CONTRIBUTING.md: Minor content updates based on reviews

- Contributing without a GitHub account
- Mention OfBorg
- nix.useSandbox -> nix.settings.sandbox
- nixpkgs-review is good for not just version updates

Co-authored-by: Rémi NICOLE <minijackson@users.noreply.github.com>

infinisil commented on CONTRIBUTING.md:
```diff
@@ -1,77 +1,193 @@
+This document is for people wanting to contribute to the implementation of Nixpkgs.
+This involves interacting with implementation changes that are proposed using [GitHub](https://github.com/) [pull requests](https://docs.github.com/pull-requests) to the [Nixpkgs](https://github.com/nixos/nixpkgs/) repository (which you're in right now).
+
+As such, a GitHub account is required, which you can sign up for [here](https://github.com/signup).
```
Good point, the Discourse link is mentioned in the [package-specific quick start guide](https://nixos.org/manual/nixpkgs/stable/#chap-quick-start), which was moved to `pkgs/README.md` in this PR. But that should definitely be lifted into the `CONTRIBUTING.md` document, I added it in 8fe2d9f34e2e3bea6366873be4905267d42378ad!

minijackson replied:
I think that will need to be discussed in the future. I've heard that patch submissions through Discourse is basically unusable, so I'm wary of recommending it.

infinisil committed:
CONTRIBUTING.md: Typos and formatting

- Fix sentence about meeting contributing standards
- pkgs -> packages
- Use emoji's because GitHub renders the :*: things weird sometimes
- Move a dot

Co-authored-by: Rémi NICOLE <minijackson@users.noreply.github.com>

infinisil commented on CONTRIBUTING.md:
```diff
@@ -16,30 +16,75 @@ This file contains general contributing information, but individual parts also h
+> Be aware that contributing implies licensing those contributions under the terms of [COPYING](./COPYING), an MIT-like license.
+
+0. Set up a local version of Nixpkgs to work with using GitHub and Git
+   1. [Fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo#forking-a-repository) the [Nixpkgs repository](https://github.com/nixos/nixpkgs/).
+   1. [Clone the forked repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo#cloning-your-forked-repository) into a local `nixpkgs` directory.
+   1. [Configure the upstream Nixpkgs repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo#configuring-git-to-sync-your-fork-with-the-upstream-repository).
```
Yeah that's intentional, I don't really mind either way, but that's an [explicitly supported GitHub Markdown](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#lists) feature (and I think effectively all markdown flavors support this)

minijackson replied:
Ok good to know. I was just confused when reading through the source code.

infinisil commented on CONTRIBUTING.md:
```diff
@@ -83,35 +83,28 @@ This section describes in some detail how changes can be made and proposed with
 
-When pull requests are made, our tooling automation bot,
-[OfBorg](https://github.com/NixOS/ofborg) will perform various checks
-to help ensure expression quality.
```
That is a good point, this is one sentence I removed because I didn't quiet know where it would fit, but I think this should work: 182d3551c5e409ee43a57e65a1160e2cf37a4b4e

infinisil commented on CONTRIBUTING.md:
```diff
@@ -205,32 +186,33 @@ gitGraph
+| [Mass rebuilds](#mass-rebuilds) | :x: | :heavy_check_mark: | :warning: Only to fix Hydra builds |
+| Critical security fixes | :heavy_check_mark: for non-mass-rebuilds | :x: | :heavy_check_mark: for mass-rebuilds |
+| Automatically merged into | `staging-next` | - | `staging` |
+| Manually merged into | - | `staging-next` | `master` |
```
Yeah I saw that too, super weird. Switching to emoji's seems to make it work correctly though, so I did that: eb758105cd3fe45a7164c03077deeaff7d0bc1da :smile:

infinisil commented on CONTRIBUTING.md:
```diff
@@ -107,56 +111,83 @@ This section describes in some detail how changes can be made and proposed with
-The oldest supported release (`YYMM`) can be found using
+When sandbox builds are enabled, Nix will setup an isolated environment for each build process. It is used to remove further hidden dependencies set by the build environment to improve reproducibility. This includes access to the network during the build outside of `fetch*` functions and files outside the Nix store. Depending on the operating system access to other resources are blocked as well ( ...
+
+Sandboxing is not enabled by default in Nix due to a small performance hit on each build. In pull requests for [nixpkgs](https://github.com/NixOS/nixpkgs/) people are asked to test builds with sandboxing enabled (see `Tested using sandboxing` in the pull request template) because in [Hydra](https://nixos.org/hydra/) sandboxing is also used.
```
This was already there before in [the manual](https://nixos.org/manual/nixpkgs/stable/#submitting-changes-tested-with-sandbox), so I don't think we should discuss it here, but yeah this definitely needs updating in the future.

infinisil commented on CONTRIBUTING.md:
````diff
@@ -107,56 +111,83 @@ This section describes in some detail how changes can be made and proposed with
+- **Globally enable sandboxing on NixOS**: add the following to `configuration.nix`
+
+  ```nix
+  nix.useSandbox = true;
````
Simple enough, made the change in fd877f5134d74bb08907e5f0db417cab2000b057

infinisil commented on CONTRIBUTING.md:
```diff
@@ -1,77 +1,193 @@
-  * If in doubt, try to contact the upstream developers for clarification.
-* `meta.mainProgram` must be set when appropriate.
-* `meta.maintainers` should be set.
+If you are updating a package’s version, you can use `nixpkgs-review` to make sure all packages that depend on the updated package still compile correctly. The `nixpkgs-review` utility can look for and build all dependencies either based on uncommitted changes with the `wip` option or specifying a GitHub pull request number.
```
This was already there before like that, but since you made the suggestion already, applying it :)

infinisil commented on CONTRIBUTING.md:
```diff
@@ -107,56 +111,83 @@ This section describes in some detail how changes can be made and proposed with
+
+#### Meets Nixpkgs contribution standards
+
+The last checkbox is fits the guidelines in this `CONTRIBUTING.md` file. The contributing document has detailed information on standards the Nix community has for commit messages, reviews, licensing of contributions you make to the project, etc... Everyone should read and understand the standards the community has for contributing before submitting a pull request.
```
Yeah, I am intending to update the pull request template with future PRs, let's not do it in this PR though

infinisil committed:
CONTRIBUTING.md: Move boot loader-specific sentence to pkgs/README.md

This needs a rewrite at some point..

infinisil commented:
I did a minor Git history cleanup by squashing all the single typo commits and other minor changes together. This is ready to be merged imo.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-08-17-documentation-team-meeting-notes-73/31853/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/this-month-in-nix-docs-4-july-august-2023/32502/4

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/frustrations-about-splicing/49607/5
